graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,377 @@
+        raise TypeError("max_components must be an int")
+    elif max_components < 1:
+        raise ValueError("max_components must be > 0")
+    elif max_components < min_components:
+        raise ValueError("max_components must be >= min_components")
+
+    if embed_kws is not None:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    """
+    Recursively clusters based on a chosen clustering algorithm
+
+    Parameters
+    ----------
+    cluster_method : str {"GMM", "KMeans", "Spherical-KMeans"}, default="GMM"
+        The clustering method chosen to apply
+    min_components : int, optional (default=1)
+        The minimum number of mixture components or clusters to consider
+        for the first split
+    max_components : int, optional (default=10)
+        The maximum number of mixture components or clusters to consider
+        at each split
+    selection_criteria : str, optional, default(=None) is the default
+        selection_criteria of chosen clustering algorithm
+        select the best model based on a certain criterion for each split
+    min_split : int, optional (default=1)
+        The minimum number of samples allowed in a leaf cluster
+    max_level : int, optional (default=20)
+        The maximum level to cluster
+    delta_criter : float or None, positive, default=None
+        The smallest difference between selection criterion values of a new
+        model and the current model that is required to accept the new model
+    likelihood_ratio : float or None, in range (0,1), default=None
+        The significance threshold of p-value for likelihood ratio test
+
+    Attributes
+    ----------
+    results_ : dict
+        Contains information about clustering results on a cluster
+        Items are:
+
+        'model' : GaussianMixture (or KMeans) object if "GMM" (or "KMeans")
+            is selected to perform clustering
+        'criter' : float
+            Bayesian (or Akaike) Information Criterion if "bic" (or "aic")
+            is chosen to select the best model
+        'n_components' : int
+            number of components or clusters
+    predictions : array_like, shape (n_samples, n_levels)
+
+    See Also
+    --------
+    graspy.cluster.AutoGMMCluster
+
+    References
+    ----------
+    .. [1]  Lyzinski, V., Tang, M., Athreya, A., Park, Y., & Priebe, C. E
+            (2016). Community detection and classification in hierarchical
+            stochastic blockmodels. IEEE Transactions on Network Science and
+            Engineering, 4(1), 13-26.
+    """
+
+    def __init__(
+        self,
+        selection_criteria=None,
+        cluster_method="GMM",
+        parent=None,
+        min_components=1,
+        max_components=10,
+        cluster_kws={},
```

Comment:
I don't think this one is documented

Reply:
not sure whether I need to add more specific input checking for cluster_kws. For example, raise an error if someone tries to set linkage="ward" when choosing "kmeans" or completely random stuff.

Reply:
too complicated, let's leave that to the user to know what the models can accept